File: 백준/Platinum/13547. 수열과 쿼리 5/수열과 쿼리 5.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>
#include <cmath>
#pragma GCC optimize ("O3")
#pragma GCC optimize ("Ofast")
#pragma GCC optimize ("unroll-loops")
#pragma GCC target("sse,sse2,sse3,ssse3,sse4,avx,avx2")

void fastIO(){
    std::cin.tie(nullptr);
    std::cout.tie(nullptr);
    std::ios_base::sync_with_stdio(false);
}

int main() {
    fastIO();
    int N;
    std::cin >> N;
    std::vector<int> A(N);
    for (int i = 0; i < N; i++) {
        std::cin >> A[i];
    }
    std::vector<int> sortedA = A;
    std::sort(sortedA.begin(), sortedA.end());
    sortedA.erase(std::unique(sortedA.begin(), sortedA.end()), sortedA.end());
    std::map<int, int> A_order;
    for (int i = 0; i < sortedA.size(); i++) {
        A_order[sortedA[i]] = i;
    }
    for (int i = 0; i < N; i++) {
        A[i] = A_order[A[i]];
    }
    int Q;
    std::cin >> Q;
    std::vector<std::pair<int, int>> queries(Q);
    for (int i = 0; i < Q; i++) {
        int s, e;
        std::cin >> s >> e;
        queries[i] = {s - 1, e - 1};
    }
    std::map<std::pair<int, int>, std::vector<int>> query_order;
    for (int i = 0; i < Q; i++) {
        query_order[queries[i]].push_back(i);
    }
    int bucket_size = sqrt(N) + 1;
    std::sort(queries.begin(), queries.end(), [bucket_size](const std::pair<int, int>& a, const std::pair<int, int>& b) {
        if (a.first / bucket_size != b.first / bucket_size) {
            return a.first / bucket_size < b.first / bucket_size;
        }
        return a.second < b.second;
    });
    std::vector<int> current_counter(N, 0);
    int current_length = 0;
    std::vector<int> answer(Q, 0);
    std::pair<int, int> last_query = {0, 0};
    for (int i = 0; i < Q; i++) {
        int s = queries[i].first;
        int e = queries[i].second;
        if (i == 0) {
            for (int j = s; j <= e; j++) {
                current_counter[A[j]]++;
                if (current_counter[A[j]] == 1) {
                    current_length++;
                }
            }
            for (int j : query_order[{s, e}]) {
                answer[j] = current_length;
            }
            last_query = {s, e};
            continue;
        }
        int s_before = last_query.first;
        int e_before = last_query.second;
        if (s_before > s) {
            for (int j = s; j < s_before; j++) {
                current_counter[A[j]]++;
                if (current_counter[A[j]] == 1) {
                    current_length++;
                }
            }
        } else if (s_before < s) {
            for (int j = s_before; j < s; j++) {
                current_counter[A[j]]--;
                if (current_counter[A[j]] == 0) {
                    current_length--;
                }
            }
        }
        if (e_before > e) {
            for (int j = e + 1; j <= e_before; j++) {
                current_counter[A[j]]--;
                if (current_counter[A[j]] == 0) {
                    current_length--;
                }
            }
        } else if (e_before < e) {
            for (int j = e_before + 1; j <= e; j++) {
                current_counter[A[j]]++;
                if (current_counter[A[j]] == 1) {
                    current_length++;
                }
            }
        }
        for (int j : query_order[{s, e}]) {
            answer[j] = current_length;
        }
        last_query = {s, e};
    }
    for (int ans : answer) {
        std::cout << ans << "\n";
    }
    return 0;
}

File: 백준/Platinum/13547. 수열과 쿼리 5/수열과 쿼리 5.py
# 14897 서로 다른 수와 쿼리 1
from collections import Counter
from sys import stdin


def input():
    return stdin.readline().strip()

def main():
    N = int(input())
    A = list(map(int, input().split()))
    # print(A)
    Q = int(input())
    queries = [tuple(map(lambda x: x - 1, map(int, input().split()))) for _ in range(Q)]
    query_order = {v: [] for v in queries}
    for i, (s, e) in enumerate(queries):
        query_order[(s, e)].append(i)
    bucket_size = int(N**0.5) + 1
    queries.sort(key=lambda x: (x[0] // bucket_size, x[1]))
    current_counter = [0 for _ in range(1000001)]
    current_length = 0
    answer = [0 for _ in range(Q)]
    last_query = [0, 0]
    for i, (s, e) in enumerate(queries):
        if i == 0:
            for j in range(s, e + 1):
                current_counter[A[j]] += 1
                if current_counter[A[j]] == 1:
                    current_length += 1
            for j in query_order[(s, e)]:
                answer[j] = current_length
            last_query = [s, e]
            continue
        s_before, e_before = last_query
        if s_before > s:
            for j in range(s, s_before):
                current_counter[A[j]] += 1
                if current_counter[A[j]] == 1:
                    current_length += 1
        elif s_before < s:
            for j in range(s_before, s):
                current_counter[A[j]] -= 1
                if current_counter[A[j]] == 0:
                    current_length -= 1
        if e_before > e:
            for j in range(e + 1, e_before + 1):
                current_counter[A[j]] -= 1
                if current_counter[A[j]] == 0:
                    current_length -= 1
        elif e_before < e:
            for j in range(e_before + 1, e + 1):
                current_counter[A[j]] += 1
                if current_counter[A[j]] == 1:
                    current_length += 1
        # print(s, e, current_counter, current_length, answer)
        for j in query_order[(s, e)]:
            answer[j] = current_length
        last_query = [s, e]
    print(*answer, sep="\n")

main()
